internal/node/controller/controller.go

```diff
@@ -135,14 +141,121 @@ func (c *BaseNodeController) FinishAppInstallation(userID string, appID, registr
 	return nil
 }
 
-func (c *BaseNodeController) GetAppByID(appID string) (*node.AppInstallation, error) {
+func (c *BaseNodeController) UpgradeApp(appID string, newAppInstallation *node.AppInstallation, newProxyRules []*node.ReverseProxyRule, version string) error {
+	nodeState, err := c.GetNodeState()
+	if err != nil {
+		return err
+	}
+
+	// Get process associated with app if it's running
+	previouslyRunning := true
+	process, err := nodeState.GetProcessForApp(appID)
+	if err != nil {
+		if _, ok := err.(node.ErrNotFound); ok {
+			// App is not running, so we can just upgrade it
+			previouslyRunning = false
+		} else {
+			return err
+		}
+	}
+
+	if previouslyRunning {
+
+		err = c.StopProcess(process.ID)
+		if err != nil {
+			return err
+		}
+		err := c.WaitForState(func(updatedState hdb.State) (bool, error) {
```

Comment:
hm this feels racy. what if there's a race condition where something else happens (say it crashes) and we never go into the upgrade state? 

I wonder if there's a better way to do this control flow.